### DISCUSSION TITLE
Miryoku VIAL json config file?

### DISCUSSION THREAD

**Discussion by PeterTheobald** [manna-harbour/miryoku#238] [Q&A] [answered]
I have a PicaChoc with VIAL-qmk. Does anyone have a Vial json config file set up for Miryoku?
If I just manually set all of the keys to Miryoku layout using the Vial app will I miss out on custom timings that make Miryoku usable?

**Comment by manna-harbour** [comment, marked as answer]
I don't maintain any Via / Vial versions of Miryoku as they would be tied to a single keyboard. You might be able to enable them on a Miryoku QMK build, but I don't support it, so you'd need to ask on their support channels for advice on what's required.

Recent versions of QMK have changed the defaults to better support home row mods, so it might be usable now without changing anything. If you can change settings, the settings used by Miryoku QMK are https://github.com/manna-harbour/miryoku_qmk/blob/miryoku/users/manna-harbour_miryoku/config.h.

**Comment by hron** [comment]
Well, I had the same issue recently because of my new keyboard. It seems Vial supports 10 layers without any issues, so I defined Miryoku there. Find my json attached. I hope you find it useful.

[miryoku.json](https://github.com/user-attachments/files/16350271/miryoku.json)

**Reply by GreyTime** [reply]
Hey there, thanks for the json. How do I get it onto the keyboard? (This might be a stupid question)

**Reply by xxxserxxx** [reply]
Do you have Vial installed? It's a GUI program, you just load the json and Vial flashes your keyboard.

**Reply by xxxserxxx** [reply]
I see the issue, though... the QMK version has a number of flags you can set to change things like the base layer, and flipping the layers. It'd be most handy to be able to convert from a QMK layout to Vial JSON.

My plan is to do the initial flash with QMK, which Vial _should_ pick up from the keyboard? I'm hoping, anyway. The Vial GUI is indispensably convenient, and I have an additional custom layer that goes back to QWERTY because Factorio in any other keyboard layout is a PITA.

**Reply by dbldippy** [reply]
@hron thanks for the json! However, I have no idea how to load it in Vial 0.7.3. Your file is not a VIA file, is it? Could you also share your `.vil` file?

Nvm. The `.json` looks like a `.vil`, so I renamed it and voila. Loads

**Comment by xxxserxxx** [comment]
Ok, oof. There's _no_ way to get a vil file from a qmk configuration; I've tried just about everything. I was going to write a script to convert a qmk keymap.json to vil, but he qmk c2json command ignores the `custom_rules.mk`, so that's out. Given that Vial is _based_ on QMK, it baffles me that they provide no way to process qmk configurations, or can parse out the keymap from a compiled `uf2` file. But I'm not going to bypass Vial, so I guess it's hand-editing for me.

What an annoying state of affairs. You'd think qmk would provide a useful generated configuration. I'm going to take a peek at some of the other implementations supported by the project; one of them must generate a usable json or something that can be parsed and transformed into a `.vil`, which is just more json, after all. If I get a tool written, I'll post back here.

**Comment by AnthonyMichaelTDM** [comment]
I have a few miroku-based vial configs for the beekeeb piantor https://github.com/AnthonyMichaelTDM/vial-configs/tree/main/piantor

**Reply by xxxserxxx** [reply]
Oh. These aren't reversed, are they? The number pad should always be on the right side, because that's how it's taught and that's where it's located on every keyboard with a number pad. I don't understand the original logic behind putting it under the left hand.

**Reply by AnthonyMichaelTDM** [reply]
they are not reversed, no

I generally agree that right-hand numpad makes more sense, but I also like having nav and mouse layers on the right hand, it's really down to personal preference honestly

**Comment by xxxserxxx** [comment]
@AnthonyMichaelTDM Dang, no Dvorak. But this is a great start; I was about to do a Dvorak version of this; I'll modify one of yours and submit a PR on your project.

**Reply by AnthonyMichaelTDM** [reply]
sounds good!
I don't have a dvorak layout since this is my personal config repo and I don't use dvorak, I might be adding a norman layout soon though as I've been playing around with it on my framework16.
that said, I'd absolutely be willing to merge such a PR

**Reply by xxxserxxx** [reply]
I have decided to eschew this, for a couple of reasons. One is that, in working with this, I realized that the Piantor Pro is surprisingly _not_ compatible with Piantor, and so things got a little crazy in the editing.

Second is that I have some layers I want (controlling tmux and the window manager) and there are some layers which have never, and will never, use that were taking up key positions (media.) Again, by the time I was done, I'd diverged from the base layout.

The biggest issue, I think, is that it was increasingly clear as I was using the layout that Miryoku was written with no regard to formal typist training -- the big tip-off was the left-handed position of the 10-key. Anybody in accounting or who's taken typing courses will have been trained with a right-hand 10-key keypad. The other issue I had with the choice to put all of the arrow keys -- and all movement keys -- on a single row, in contravention to the traditional inverted T layout. The actual project has fixes for both of these, which is fantastic, but I couldn't think of an easy way to see how the project lays out the keyboard with the INVERTED and FLIP flags so I put them where I'm used to them. Finally, I have a layer dedicated to QWERTY, for games, because it's easier to toggle to it and use the game's official layout that re-configuring every game -- especially when the game is Factorio and every single key and modifier-key combo is bound to some function. By now, I was so far off "standard", I replaced the symbols layer with the positioning I was used to, and there's no point anymore to calling my monstrosity "Miryoku".

The real value I got from Miryoku was the focus on using the thumbs for layers. The reason I was looking at Miryoku in the first place is because I had all of the layer switching under the home row; I'm a fast typer and I just could not get the tap-dance/macro/combo configurations set up such that I wasn't either accidentally triggering a layer switch, or getting frustrated at not being able to switch layers fast enough to type punctuation. Miryoku's layout is far more simple and straightforward. It'll take a while adjusting to having the modifiers under the home row, but I do suspect moving the layer controls to the thumbs is going to improve my experience in the long run. I should have set up my piantor that way in the first place. The only real downside I can foresee is that my layered homerow layout works on any keyboard with [kanata](https://github.com/jtroo/kanata), even if they don't have extra thumb keys, which makes it _far_ easier to switch between keyboards. We'll see.

Anyway, my apologies, but my configuration no longer much resembles Miryoku; I can still PR it, if you like, but it wouldn't match the documentation.
